splash-reward-distributor/src/indexer.rs

```diff
@@ -0,0 +1,641 @@
+use std::hash::Hash;
+use std::path::Path;
+use std::{
+    fmt::{Debug, Display},
+    sync::Arc,
+};
+
+use bloom_offchain::execution_engine::bundled::Bundled;
+use const_format::formatcp;
+use log::trace;
+use rocksdb::{Options, TransactionDB, TransactionDBOptions};
+use serde::Deserialize;
+use serde::{de::DeserializeOwned, Serialize};
+use spectrum_offchain::domain::Stable;
+use spectrum_offchain::domain::{
+    event::{AnyMod, Confirmed, Predicted, Traced},
+    EntitySnapshot,
+};
+use tokio::task::spawn_blocking;
+
+use crate::onchain::{
+    auth_manager::AuthManager, buffer_wallet::BufferWallet, harvest_order::HarvestOrder, smart_farm::Gauge,
+};
+
+#[async_trait::async_trait]
+pub trait OnChainIndex<Bearer>
+where
+    Bearer: Serialize + DeserializeOwned,
+{
+    async fn read<T>(&self, id: T::StableId) -> Option<AnyMod<Bundled<T, Bearer>>>
+    where
+        T: unique_ids::UniqueId + EntitySnapshot + Send + Serialize + DeserializeOwned + 'static,
+        T::StableId: Serialize + DeserializeOwned + 'static;
+
+    async fn write_predicted<T>(&self, entity: Traced<Predicted<Bundled<T, Bearer>>>)
+    where
+        T: unique_ids::UniqueId + EntitySnapshot + Send + Clone + Serialize + 'static,
+        T::StableId: Serialize + DeserializeOwned + 'static;
+
+    async fn write_confirmed<T>(&self, entity: Traced<Confirmed<Bundled<T, Bearer>>>)
+    where
+        T: unique_ids::UniqueId + EntitySnapshot + Send + Clone + Serialize + 'static,
+        T::StableId: Serialize + DeserializeOwned + 'static;
+
+    /// Deletes latest version of the entity and returns the previous version if it exists.
+    async fn remove<T>(&self, id: T::StableId, version: T::Version) -> Option<T::Version>
```

Comment:
Made slight change to API here. Caller needs to provide both the StableID and version of the entity to be removed. The `version` parameter is used within an assertion to make sure the latest version is being deleted.